### ISSUE TITLE
Investigate bi-stable BitArrayBoolArrayCtor benchmark

### ISSUE THREAD

**Issue by adamsitnik** [dotnet/performance#697]
`new BitArray(bool[])` is 2x slower compared to 2.2

https://github.com/dotnet/performance/blob/ce894eec66c6f83b227db37f9833b2116ba54c48/src/benchmarks/micro/corefx/System.Collections/Perf.BitArray.cs#L38-L42

Repro

```cmd
git clone https://github.com/dotnet/performance.git
cd performance
# if you don't have cli installed and want python script to download the latest cli for you
py .\scripts\benchmarks_ci.py -f netcoreapp2.2 netcoreapp3.0 --filter *BitArrayBoolArrayCtor
# if you do
dotnet run -p .\src\benchmarks\micro\MicroBenchmarks.csproj -c Release -f netcoreapp2.2 --filter *BitArrayBoolArrayCtor --runtimes netcoreapp2.2 netcoreapp3.0
```

```ini
BenchmarkDotNet=v0.11.3.1003-nightly, OS=Windows 10.0.18362
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview8-013262
   [Host]     : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT
   Job-BYJCMJ : .NET Core 2.2.6 (CoreCLR 4.6.27817.03, CoreFX 4.6.27818.02), 64bit RyuJIT
   Job-JSSCYO : .NET Core 3.0.0-preview8-27916-02 (CoreCLR 4.700.19.36302, CoreFX 4.700.19.36514), 64bit RyuJIT
```
|                Method |     Toolchain | Size |       Mean | Ratio | Allocated Memory/Op |
|---------------------- |-------------- |----- |-----------:|------:|--------------------:|
| BitArrayBoolArrayCtor | netcoreapp2.2 |  512 |   657.1 ns |  1.00 |               128 B |
| BitArrayBoolArrayCtor | netcoreapp3.0 |  512 | 1,468.7 ns |  2.23 |               120 B |

/cc @danmosemsft

**Comment by adamsitnik** [comment]
@billwert I've transferred this issue from CoreFX to performance repo because it's most probably a benchmarking issue or some hardware phenomenon. Identifying the source of the issue might not be worth the struggle per se, but it can give us some insight for similar issues in the future (like our unstable Linux runs). 

It's a great low-leve profiling exercise, I am going to mark it as up-for-grabs so maybe somebody will pick it up.

**Comment by Paxxi** [comment]
I've managed to reproduce it at will on Windows. The first run I do after building Microbenchmarks.csproj shows the regression and any subsequent builds perform as they should. Cleaning the build and running again shows the exact same behavior that the first run is slow.

<details><summary>Assembly</summary>
first run

```assembly
; System.Collections.BitArray..ctor(Boolean[])
       push    rdi
       push    rsi
       push    rbx
       sub     rsp,20h
       mov     rdi,rcx
       mov     rsi,rdx
       test    rsi,rsi
       je      System_Collections+0x1f3b3
       mov     ebx,dword ptr [rsi+8]
       mov     ecx,ebx
       add     ecx,1Fh
       shr     ecx,5
       movsxd  rcx,ecx
       call    qword ptr [System_Collections+0x10b0
       lea     rcx,[rdi+8]
       mov     rdx,rax
       call    qword ptr [System_Collections+0x1018
       mov     dword ptr [rdi+10h],ebx
       xor     eax,eax
       test    ebx,ebx
       jle     System_Collections+0x1f3a6
       movsxd  rcx,eax
       cmp     byte ptr [rsi+rcx+10h],0
       je      System_Collections+0x1f3a0
       mov     ecx,eax
       shr     ecx,5
       mov     edx,eax
       and     edx,1Fh
       mov     r8,qword ptr [rdi+8]
       cmp     ecx,dword ptr [r8+8]
       jae     System_Collections+0x1f3db
       movsxd  rcx,ecx
       lea     r8,[r8+rcx*4+10h]
       mov     r9d,1
       mov     ecx,edx
       shl     r9d,cl
       or      dword ptr [r8],r9d
       inc     eax
       cmp     ebx,eax
       jg      System_Collections+0x1f36c
       xor     eax,eax
       mov     dword ptr [rdi+14h],eax
       add     rsp,20h
       pop     rbx
       pop     rsi
       pop     rdi
       ret
       call    qword ptr [System_Collections+0x10f8
       mov     rsi,rax
       mov     ecx,805h
       call    System_Collections+0x4703
       mov     rdx,rax
       mov     rcx,rsi
       call    qword ptr [System_Collections+0x2430
       mov     rcx,rsi
       call    System_Collections+0x4710
       int     3
       call    System_Collections+0x4716
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       int     3
       push    rdi
       push    rsi
       push    rbx
       sub     rsp,30h
       mov     rsi,rcx
       mov     rdi,rdx
; Total bytes of code 205
```
subsequent runs
```assembly
; System.Collections.BitArray..ctor(Boolean[])
       push    rdi
       push    rsi
       push    rbx
       sub     rsp,20h
       mov     rsi,rcx
       mov     rdi,rdx
       test    rdi,rdi
       je      System_Collections+0x1e65e
       mov     ebx,dword ptr [rdi+8]
       mov     ecx,ebx
       test    ecx,ecx
       jg      System_Collections+0x1e5d3
       xor     ecx,ecx
       jmp     System_Collections+0x1e5e5
       dec     ecx
       mov     eax,ecx
       sar     eax,1Fh
       and     eax,1Fh
       add     eax,ecx
       sar     eax,5
       lea     ecx,[rax+1]
       call    qword ptr [System_Collections+0x10d0
       lea     rcx,[rsi+8]
       mov     rdx,rax
       call    qword ptr [System_Collections+0x1020
       mov     dword ptr [rsi+18h],ebx
       xor     eax,eax
       test    ebx,ebx
       jle     System_Collections+0x1e651
       movsxd  rcx,eax
       cmp     byte ptr [rdi+rcx+10h],0
       je      System_Collections+0x1e64b
       mov     rcx,qword ptr [rsi+8]
       mov     edx,eax
       sar     edx,1Fh
       and     edx,1Fh
       add     edx,eax
       sar     edx,5
       cmp     edx,dword ptr [rcx+8]
       jae     System_Collections+0x1e686
       movsxd  rdx,edx
       lea     rdx,[rcx+rdx*4+10h]
       mov     ecx,eax
       sar     ecx,1Fh
       and     ecx,1Fh
       add     ecx,eax
       and     ecx,0FFFFFFE0h
       mov     r8d,eax
       sub     r8d,ecx
       mov     r9d,1
       mov     ecx,r8d
       shl     r9d,cl
       or      dword ptr [rdx],r9d
       inc     eax
       cmp     ebx,eax
       jg      System_Collections+0x1e601
       xor     eax,eax
       mov     dword ptr [rsi+1Ch],eax
       add     rsp,20h
       pop     rbx
       pop     rsi
       pop     rdi
       ret
       call    qword ptr [System_Collections+0x10e8
       mov     rsi,rax
       mov     ecx,805h
       call    System_Collections+0x47d9
       mov     rdx,rax
       mov     rcx,rsi
       call    qword ptr [System_Collections+0x2880
       mov     rcx,rsi
       call    System_Collections+0x47e6
       int     3
       call    System_Collections+0x47ec
       int     3
       int     3
       int     3
       int     3
       int     3
       push    rdi
       push    rsi
       push    rbx
       sub     rsp,30h
       mov     rsi,rcx
       mov     rdi,rdx
       test    rdi,rdi
       je      System_Collections+0x1e6eb
       mov     ebx,dword ptr [rdi+8]
       cmp     ebx,3FFFFFFh
; Total bytes of code 251
```
</details>
For some reason the first run has worse assembly code generated for it than the other runs.
I've seen no such difference in the assembly generated for 2.2

**Comment by danmosemsft** [comment]
> For some reason the first run has worse assembly code generated for it than the other runs. I've seen no such difference in the assembly generated for 2.2

Sounds like tiered jitting again?

**Comment by Paxxi** [comment]
That's a good point. I will spend some more time on it today and digging deeper into it. Have a lot to learn both about BDT and the JIT 😀

**Comment by danmosemsft** [comment]
cc @kouvel

**Comment by billwert** [comment]
given that it sounds like it is separate process invocations ("The first run I do after building Microbenchmarks.csproj shows the regression and any subsequent builds perform as they should.") that's not going to be an impact of tiered jitting.

**Comment by Paxxi** [comment]
My machine rebooted and now it's not "stable" anymore, but still reproducible.

I don't believe it has anything to do with tiered jitting because of the following
- None of the generated projects specify tiered jit
- The published rules mention 30 invocations to trigger optimised jit and there weren't 30 calls

What I see now is that randomly one of the benchmarked methods is optimised when loaded and the other isn't. Specifying `MinWarmupCount(20)` reliable causes both methods to be optimised on my machine. Anything less than 20 is not reliable on my machine. It's easy to see when the optimisation happens during the warmup as the execution time drops from ~1800 to ~700.

Looking at the `Microsoft-Windows-DotNETRuntime/Method/LoadVerbose` event and filtering on `BitArrayBoolArrayCtor` it shows that one method has MethodLoadFlags=Jitted and the other has MethodLoadFlags=392 and then there's another event where it's 520 and ReJITID=1.

The values 392 and 520 aren't documented but looking at the code in CoreClr these correspond to this enum left shifted 7 bits.
```csharp
    enum class JitOptimizationTier : UINT8
    {
        Unknown, // to identify older runtimes that would send this value
        MinOptJitted,
        Optimized,
        QuickJitted,
        OptimizedTier1,

        Count
    };
```
Plugging this into calculator gives 392>>7=3(QuickJitted) and 520>>7=4(OptimizedTier1).

In summary I'm fairly certain about what happens and the easy solution is to use `MinWarmupCount`. Why it happens is still unknown to me and I'm going to try to figure it out but it's likely time consuming 😄 Is there anyone on the JIT/MethodLoader teams available to provide some insights into why this happens?

@adamsitnik What do you say about the currently proposed solution?

**Comment by Paxxi** [comment]
This is fun, results are stable during a session of a couple of hours but so far have behaved differently every day I've looked at it.

`MinWarmupCount(20)` is no longer enough to trigger a rejit.

Best thing might be to manually inline the method. Using `MethodImpl(AggressiveInline)` didn't give as good results

**Comment by billwert** [comment]
I wonder if this has something to do with image layout. Changing between builds might mean that the assemblies are loading in different relative places, so we emit different size jumps (or something of that nature.) We've seen things like that in the past. I will spend some time trying to see if I can catch a good and bad.

Regarding warmup count, each iteration does several invocations of the actual benchmark. On my machine this test does ~500K invocations per iteration. I'd expect the test method to be tiered up before the first warmup invocation is completed. If I specify a `--warmupCount 1 --invocationCount 1`, I see both the `BitArray()` ctor as well as the `BitArrayBoolArrayCtor` test method being rejit to Tier 1. I'd like us to have the actual root cause for the noise before we start trying various things which may or may not fix it.

**Comment by billwert** [comment]
Here's details from the trace:
Method rejits:

Event Name | Time MSec | Process Name | Rest
-- | -- | -- | --
Microsoft-Windows-DotNETRuntime/Method/LoadVerbose | 701.327 | dotnet (12888) | ThreadID="12,552" ProcessorNumber="6" MethodID="140,727,119,579,512" ModuleID="140,727,119,298,544" MethodStartAddress="0x7ffd96062430" MethodSize="49" MethodToken="100,668,325" MethodFlags="Jitted" MethodNamespace="System.Collections.Tests.Perf_BitArray" MethodName="BitArrayBoolArrayCtor" MethodSignature="instance class System.Collections.BitArray  ()" ClrInstanceID="8" ReJITID="1" OptimizationTier="OptimizedTier1"
Microsoft-Windows-DotNETRuntime/Method/LoadVerbose | 701.674 | dotnet (12888) | ThreadID="12,552" ProcessorNumber="6" MethodID="140,727,119,619,832" ModuleID="140,727,119,206,832" MethodStartAddress="0x7ffd96062480" MethodSize="203" MethodToken="100,663,346" MethodFlags="Jitted" MethodNamespace="System.Collections.BitArray" MethodName=".ctor" MethodSignature="instance void  (bool[])" ClrInstanceID="8" ReJITID="1" OptimizationTier="OptimizedTier1"



That puts it in what is the 3rd pilot iteration on my machine:

Event Name | Time MSec | Process Name | Rest
-- | -- | -- | --
BenchmarkDotNet.EngineEventSource/WorkloadPilot/Start | 501.907 | dotnet (12888) | ThreadID="12,460" ProcessorNumber="2" totalOperations="551,696"
BenchmarkDotNet.EngineEventSource/WorkloadPilot/Stop | 740.617 | dotnet (12888) | ThreadID="12,460" ProcessorNumber="2" DURATION_MSEC="238.710" totalOperations="551,696"

```
WorkloadPilot    1: 16 op, 16200.00 ns, 1.0125 us/op
WorkloadPilot    2: 246928 op, 111897400.00 ns, 453.1580 ns/op
WorkloadPilot    3: 551696 op, 238697700.00 ns, 432.6616 ns/op
```


It is still true that the methods are executed several hundred thousand times before they are re-jit. Perhaps it's the case that when conditions align perfectly, the background thread never gets a chance to promote the methods in question to tier 1. That seems incredibly unlikely to me, though.

**Comment by Paxxi** [comment]
What I've seen the generated assembly is deterministic in that the bad and good versions always look the same. It wouldn't be the first time in this issue that I noticed patterns that aren't really stable though 😀

Thanks for the explanation 😀

**Comment by Paxxi** [comment]
Here's two runs almost back to back, no other runs between them, the first showed bad performance for 2.2 and the second bad performance for 3.0.
https://1drv.ms/u/s!AkvQauypUh5JjJwL5ErYklCZsThPGQ
https://1drv.ms/u/s!AkvQauypUh5JjJwM3ovRWDb9zBO0mA

Comparing the InliningSuccessful events shows that one TFM had the DivRem method inlined but not the other. It looks like this happens in both runs.

**Comment by billwert** [comment]
@Paxxi neat! thanks! Since you didn't share the ETLs, can you share the InliningFailed events so we can see what the reason was?

**Comment by Paxxi** [comment]
The etls Should be in the date folder, fairly deep in the folder structure . I zipped the output from BDN
